### ISSUE TITLE
Jack Driver does not handle drivers/ports correctly

### ISSUE THREAD

**Issue by kunstmusik** [csound/csound#692]
I receieved a question about JACK device detection on the Blue mailing list
and it lead me to wonder if Csound is doing the correct thing for Jack
devices in rtjack.c.  The following is the user's "jack_lsp -t -p"
output:

```
system:capture_1
    properties: output,physical,terminal,
    32 bit float mono audio
system:capture_2
    properties: output,physical,terminal,
    32 bit float mono audio
system:playback_1
    properties: input,physical,terminal,
    32 bit float mono audio
system:playback_2
    properties: input,physical,terminal,
    32 bit float mono audio
system:midi_capture_1
    properties: output,physical,terminal,
    8 bit raw midi
system:midi_playback_1
    properties: input,physical,terminal,
    8 bit raw midi
PulseAudio JACK Sink:front-left
    properties: output,terminal,
    32 bit float mono audio
PulseAudio JACK Sink:front-right
    properties: output,terminal,
    32 bit float mono audio
PulseAudio JACK Source:front-left
    properties: input,terminal,
    32 bit float mono audio
PulseAudio JACK Source:front-right
    properties: input,terminal,
    32 bit float mono audio
```

and this is the list of devices shown by "csound -+rtaudio=jack --devices":

```
virtual_keyboard real time MIDI plugin for Csound
0dBFS level = 32768.0
Csound version 6.07 (double samples) May  6 2016
libsndfile-1.0.25
rtaudio: JACK module enabled
rtmidi: ALSA Raw MIDI module enabled
3 audio input devices
 0: adc:system:capture_ (system:capture_)
 1: adc:PulseAudio JACK Sink:front-left (PulseAudio JACK Sink:front-left)
 2: adc:PulseAudio JACK Sink:front-right (PulseAudio JACK Sink:front-right)
3 audio output devices
 0: dac:system:playback_ (system:playback_)
 1: dac:PulseAudio JACK Source:front-left (PulseAudio JACK Source:front-left)
 2: dac:PulseAudio JACK Source:front-right (PulseAudio JACK Source:front-right)
csound->FileOpen2 failed:: No such file or directory
csound->FileOpen2 failed:: No such file or directory
WARNING: could not open library
'/usr/local/lib/csound/plugins-6.0/libstkops.so' (libstk.so.0: cannot
open shared object file: No such file or directory)
end of score.           overall amps:      0.0
       overall samples out of range:        0
0 errors in performance
Elapsed time at end of performance: real: 0.366s, CPU: 0.118s
```

The issue here is the "PulseAudio JACK Source:front-right" and
front-left ports.  It does not use numbers in its name, which caused a
parsing issue in Blue as I had assumed that would be the case.  In
Csound, it shows as two separate devices, rather than a single device
with 2 channels.

jlucas on the csound-dev list noted that a similar issue happened with a Firewire device and the only way they could use Csound was to modify Csound's code and recompile.

**Comment by vlazzarini** [comment]
So, the code that lists the ports does this:

1. Checks the port name for digits at the end.
2. if it finds these, then it compares the name of this port with the previous,
    if they match, it is it is made to be the same 'logical' output, and the digits
   are the channel indices.
3. if there are no digits, each port is a separate 'logical' output.

I think this makes sense to a certain extent, because there is no explicit way to
match output and port to connect automatically. In the case of the non-numeric
port endings, it is hard to parse for the general case. So we have to find a way
to explicitly name ports to connect the channels. One idea is this

-odac:<first_channel_port>;<second_channel_port>[;...]

in other words a ;-separated list. Colons can't be used as they are already
in port names.

-odac:PulseAudio JACK Source:front-left;PulseAudio JACK Source:front-right;

for stereo output.

How does that sound?

**Comment by gogins** [comment]
You could look at how the Jacko opcodes do this.

Regards,
Mike


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by vlazzarini** [comment]
I know how to do this; I need to find the best way to allow users to express the wish to auto connect an out/in channel to a given port.

**Comment by vlazzarini** [comment]
Ok; I've modified the scheme as this:
1) users can either give numbers or names to -odac
2) if using names, the old scheme works as before (:clientname:portname_) with the numeric
value omitted when selecting the port channel for autoconnection.
3) if using numbers, the number given will be the base for auto connection, so for a base port number N we will have the connections as N+0, N+1, ..., N+nchnls-1.
4) Ports are listed separately and numbered (by the listDevices() API callback).
 0: dac0 (system:playback_1)
 1: dac1 (system:playback_2)

So autoconnection can be done for any number of adjacent ports, indepedendent of how
they are named. There is no way to introduce arbitrary autoconnection with -odac / -iadc
without going into a complicated parsing/naming convention. We can think of adding
some other means of autoconnection from a text file or even in the orchestra. But for now,
I am trying to address the bug as defined in this ticket.

**Comment by jpffitch** [comment]
I realise this is boring, but ' think we need some manual words for this

**Comment by jpffitch** [comment]
In manual overview/realtimelinux.xml

**Comment by tjingboem** [comment]
Since a few years the Jack system is available for Windows and MacOSX as well:
http://jackaudio.org/downloads/
Should there be a quote on where to find info about Jack be added to JackRealTimeMac.html and RealTimeWindows.html?

**Comment by vlazzarini** [comment]
I will add man info once we are happy with these changes.

**Comment by vlazzarini** [comment]
I've documented this and moved all the jack IO info to its own separate page.
